### PR TITLE
Update Generator.Rewriter to Cecil 0.10

### PR DESCRIPTION
**Pull request by Nihlus**
With the release of Mono 5.0 and the emergence of Roslyn, a new debugging format in the form of portable debug files (.pdb portable) came up. Cecil 0.9 does not support this format, and as such the rewriter does not function with newer toolchains or compilers.

This pull request updates the Cecil version that the rewriter uses to 0.10, patches breaking changes due to this update, removes some obsolete Cecil code and fixes a number of smaller issues.

Two important notes:

- [Cecil 0.10 no longer supports .NET 2.0 or the Compact Framework](http://cecil.pe/post/149243207656/mono-cecil-010-beta-1)
- Cecil 0.10 is a beta release

Summarized highlights:

- [Feature] Supports the new portable and embedded PDB formats through Cecil
- [Fix] Tighter checking that the generated variables are the correct ones when emitting IL
- [Fix] Changed input options type to prevent multiple enumeration
- [Fix] Disposes the assembly that is to be rewritten after using it
- [Code cleanup] Added explicit static modifier to some methods which could have it

### CODE REVIEW

File: src/Generator.Rewrite/GeneratedVariableIdentifier.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+using System;
+using Mono.Cecil.Cil;
+
+namespace OpenTK.Rewrite
+{
+    /// <summary>
+    /// Acts as a unique identifier for a generated named variable that can be passed between methods. Replaces uses of 
+    /// variable names from Mono.Cecil.
+    /// </summary>
+    public class GeneratedVariableIdentifier
```
Comment (varon) [inline src/Generator.Rewrite/GeneratedVariableIdentifier.cs]: can we mark this as sealed and/or internal?

File: src/Generator.Rewrite/Program.cs [hunk 2]
```diff
@@ -61,97 +60,91 @@ static void Main(string[] args)
 
         static bool dllimport;
 
-        void Rewrite(string file, string keyfile, IEnumerable<string> options)
+        void Rewrite(string file, string keyfile, List<string> options)
```
Comment (varon) [inline src/Generator.Rewrite/Program.cs]: If this signature was changed, we definitely want to change it to `IReadOnlyList<T>` or `IReadOnlyCollection<T>`. If neither is available, due to this running on ancient, decrepit versions of .Net, keep the original signature and perform the .ToList() inside the scope of the method to ensure that the enumration of the parameter only takes place once.

File: src/Generator.Rewrite/Program.cs [hunk 3]
```diff
@@ -61,97 +60,91 @@ static void Main(string[] args)
...
 
-            // Save rewritten assembly
-            assembly.Write(file, write_params);
+                // Save rewritten assembly
+                assembly.Write(write_params);   
+            }
         }
 
-        void Rewrite(TypeDefinition type, IEnumerable<string> options)
+        void Rewrite(TypeDefinition type, List<string> options)
```
Comment (varon) [inline src/Generator.Rewrite/Program.cs]: Same thing as the other signature.

File: src/Generator.Rewrite/Program.cs [hunk 4]
```diff
@@ -226,7 +219,7 @@ int GetSlot(MethodDefinition signature)
             }
         }
 
-        void RemoveNativeSignatures(TypeDefinition type, List<MethodDefinition> methods)
+        static void RemoveNativeSignatures(TypeDefinition type, List<MethodDefinition> methods)
```
Comment (varon) [inline src/Generator.Rewrite/Program.cs]: Parameter should probably be IEnumerable for consistency here.

File: src/Generator.Rewrite/Program.cs [hunk 5]
```diff
@@ -171,12 +164,12 @@ void Rewrite(TypeDefinition type, IEnumerable<string> options)
...
                 var rewritten = new CustomAttribute(rewritten_constructor);
                 rewritten.ConstructorArguments.Add(new CustomAttributeArgument(
-                    type.Module.Import(mscorlib.MainModule.GetType("System.Boolean")), true));
+                    type.Module.ImportReference(mscorlib.MainModule.GetType("System.Boolean")), true));
                 type.Module.Assembly.CustomAttributes.Add(rewritten);
             }
         }
 
-        int GetSlot(MethodDefinition signature)
+        static int GetSlot(MethodDefinition signature)
```
Comment (varon) [inline src/Generator.Rewrite/Program.cs]: Nice additions of static to the methods.

File: src/Generator.Rewrite/Program.cs [hunk 6]
```diff
@@ -255,7 +248,7 @@ void RemoveSupportingAttributes(TypeDefinition type)
 
         // Create body for method
         static void ProcessMethod(MethodDefinition wrapper, MethodDefinition native, int slot,
-                                  FieldDefinition entry_points, IEnumerable<string> options)
+                                  FieldDefinition entry_points, List<string> options)
```
Comment (varon) [inline src/Generator.Rewrite/Program.cs]: Follow the "IEnumerable ToList internally" pattern again here if multiple enumeration is required.

File: src/Generator.Rewrite/Program.cs [hunk 7]
```diff
@@ -521,28 +515,29 @@ private static void EmitReturnTypeWrapper(MethodDefinition wrapper, MethodDefini
...
+        static void EmitParameterEpilogues(MethodDefinition wrapper, MethodDefinition native, MethodBody body, ILProcessor il, 
+            List<GeneratedVariableIdentifier> generatedVariables)
         {
-            foreach (var p in wrapper.Parameters)
+            foreach (var p in wrapper.Parameters) 
             {
                 if (p.ParameterType.Name == "StringBuilder")
                 {
-                    EmitStringBuilderEpilogue(wrapper, native, p, body, il);
+                    EmitStringBuilderEpilogue(wrapper, native, p, body, il, generatedVariables.FirstOrDefault(v => v.Name == p.Name + "_sb_ptr" && v.Body == body && body.Variables.Contains(v.Definition)));
```
Comment (varon) [inline src/Generator.Rewrite/Program.cs]: Can we refactor the multiple occurances of this logic into a helper method to find for a given name?

File: src/Generator.Rewrite/GeneratedVariableIdentifier.cs (hunk 1 above)
Comment (Nihlus) [inline src/Generator.Rewrite/GeneratedVariableIdentifier.cs]: Yep, reasonable.

File: src/Generator.Rewrite/Program.cs (hunk 7 above)
Comment (Nihlus) [inline src/Generator.Rewrite/Program.cs]: That can be done. I'll pull it out.

File: src/Generator.Rewrite/Program.cs (hunk 4 above)
Comment (Nihlus) [inline src/Generator.Rewrite/Program.cs]: Agreed.

**Comment by varon** [comment]
Thanks for the super hasty update!

Before I can merge, we need to run a test of this to confirm it's actually working.
Even better would be to compare the IL after rewrite before and after this change.

Which works best for you?

**Comment by Nihlus** [comment]
I've used the generated assemblies in a nontrivial application, and at first glance there are no errors. Everything renders as expected and works as before.

Comparing the IL sounds like a good way of being really sure - I'll see what I can do about that tomorrow.

**Comment by Nihlus** [comment]
I've been reading through the IL of the original assembly (taken from Nuget) and the new assembly. The generated IL is identical in both versions, save for the code which has not undergone rewriting. That IL differs in terms of attribute ordering, some additional assembly information, content ordering, etc - all expected when using different compilers at different times.

This is a screenshot of the start of the IL block for GL.ProgramUniformMatrix3. The line numbers differ for the reasons listed above, but the synchronization points are clearly visible with identical IL in between. This pattern is identical across other rewritten methods.

image

I've attached both assemblies if you'd like to decompile and compare yourself, as well as the IL I decompiled and compared.
[Assemblies+IL.zip](https://github.com/opentk/opentk/files/1037911/Assemblies.IL.zip)

**Comment by varon** [comment]
Awesome. I really appreciate your rigour and well-thought out approach in tackling this.

**Pull request merged**